### DISCUSSION TITLE
Roundcube/Dovecot oauth stopped working with Authelia v4.38.6: "Introspection failed: No username returned"

### DISCUSSION THREAD

**Discussion by sphakka** [authelia/authelia#7048] [Troubleshooting] [answered]
I had previously Authelia v4.37.5 (bare-metal)  providing OIDC auth to Roundcube (v1.6.6) + Dovecot (v2.3.21) on Debian 12. Now, with Authelia v4.38.6, Roundcube, after a successful user authentication, fails at IMAP login with Dovecot, who complains: **"oauth2 failed: Introspection failed: No username returned".** 

Hopefully, it's just a misconfiguration issue, though I have no idea where to look!  Can anybody please help?

## Configuration

Here are the relevant snippets.

### Authelia

```YAML
identity_providers:
  oidc:
    hmac_secret: ***
    enable_client_debug_messages: true
    jwks:
      - key: {{ secret "/etc/authelia/secrets/oidc.private.pem" | mindent 10 "|" | msquote }}
    clients:
      -
        client_id: authelia-roundcube-oidc
        client_name: 'Authelia OIDC for local Roundcube webmail instance'
        client_secret: ***
        consent_mode: implicit
        scopes:
          - openid
          - email
          - profile
        redirect_uris:
          - https://example.com/index.php/login/oauth
        token_endpoint_auth_method: 'client_secret_post'
```
### Roundcube

```PHP
$config['oauth_provider'] = 'generic';
$config['oauth_provider_name'] = 'Authelia OIDC';
$config['oauth_client_id'] = 'authelia-roundcube-oidc';
$config['oauth_client_secret'] = '***';
$config['oauth_auth_uri'] = 'https://example.com/api/oidc/authorization';
$config['oauth_token_uri'] = 'https://example.com/api/oidc/token';
$config['oauth_identity_uri'] = 'https://example.com/api/oidc/userinfo';
$config['oauth_identity_fields'] = ['email'];
$config['oauth_scope'] = 'email openid profile';
$config['oauth_login_redirect'] = true;
```

### Dovecot

(See https://doc.dovecot.org/configuration_manual/authentication/oauth2/)

`dovecot.conf`:

```shell
auth_mechanisms = plain login xoauth2 oauthbearer
passdb {
    driver = oauth2
    mechanisms = xoauth2 oauthbearer
    args = /etc/dovecot/dovecot-oauth2.conf.ext
}

```

`dovecot-oauth2.conf.ext`:

```shell
introspection_mode = post
introspection_url = https://authelia-roundcube-oidc:***@example.com/api/oidc/introspection
username_attribute = username
debug = yes
```

### Authelia

```
time="2024-03-30T19:12:15+01:00" level=debug msg="Authorization Request with id '4f16f90d-989d-4194-827e-46f39e8f39ba' on client with id 'authelia-roundcube-oidc' is being processed" method=GET path=/api/oidc/authorization remote_ip="*redacted*"
time="2024-03-30T19:12:15+01:00" level=debug msg="Authorization Request with id '4f16f90d-989d-4194-827e-46f39e8f39ba' on client with id 'authelia-roundcube-oidc' was successfully processed, proceeding to build Authorization Response" method=GET path=/api/oidc/authorization remote_ip="*redacted*"
time="2024-03-30T19:12:15+01:00" level=debug msg="Access Request with id '4f16f90d-989d-4194-827e-46f39e8f39ba' on client with id 'authelia-roundcube-oidc' is being processed" method=POST path=/api/oidc/token remote_ip="*redacted*"
time="2024-03-30T19:12:15+01:00" level=debug msg="Access Request with id '4f16f90d-989d-4194-827e-46f39e8f39ba' on client with id 'authelia-roundcube-oidc' has successfully been processed" method=POST path=/api/oidc/token remote_ip="*redacted*"
time="2024-03-30T19:12:16+01:00" level=debug msg="Introspection Request with id '5679610b-b75c-4d66-9fe4-3714d8244ae2' is being processed" method=POST path=/api/oidc/introspection remote_ip="*redacted*"
time="2024-03-30T19:12:16+01:00" level=debug msg="Introspection Request with id '5679610b-b75c-4d66-9fe4-3714d8244ae2' was processed successfully" method=POST path=/api/oidc/introspection remote_ip="*redacted*"
```

### Roundcube

```
[2024-Mar-30 18:12:15 +0000]: <vluarv91> [9A52] Connecting to localhost:143...
[2024-Mar-30 18:12:16 +0000]: <vluarv91> [9A52] S: * OK [CAPABILITY IMAP4rev1 SASL-IR LOGIN-REFERRALS ID ENABLE IDLE LITERAL+ STARTTLS AUTH=PLAIN AUTH=LOGIN AUTH=XOAUTH2 AUTH=OAUTHBEARER] Dovecot (Debian) ready.
[2024-Mar-30 18:12:16 +0000]: <vluarv91> [9A52] C: A0001 AUTHENTICATE ****** [202]
[2024-Mar-30 18:12:18 +0000]: <vluarv91> [9A52] S: + eyJzdGF0dXMiOiI0MDEiLCJzY2hlbWVzIjoiYmVhcmVyIiwic2NvcGUiOiJtYWlsIn0=
[2024-Mar-30 18:12:18 +0000]: <vluarv91> [9A52] C: ****** [-2]
[2024-Mar-30 18:12:18 +0000]: <vluarv91> [9A52] S: A0001 NO [UNAVAILABLE] Temporary authentication failure. [***:2024-03-30 18:12:18]
[2024-Mar-30 18:12:18 +0000]: <vluarv91> IMAP Error: Login failed for user@example.com against localhost from *redacted*. AUTHENTICATE XOAUTH2: A0001 NO [UNAVAILABLE] Temporary authentication failure. [***:2024-03-30 18:12:18] in /usr/share/roundcube/program/lib/Roundcube/rcube_imap.php on line 211 (GET /index.php/login/oauth?code=authelia_ac_tRk8qAayfBxbssY_1XII2vYTFUSqkDpnBEUW7dwhSz8.JCsxtZZkulh1YP1ZAksoF0DWB-mWHZ7YbBfXkG3hG20&iss=https%3A%2F%2Fauth.example.com&scope=email+openid+profile&state=cOK61wdUIYDQ)
```

### Dovecot

```
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth client connected (pid=509949)
Mar 30 19:12:16 myvps dovecot: auth: Debug: client in: AUTH#0111#011XOAUTH2#011service=imap#011secured#011session=9dBBsuQUPM1/AAAB#011lip=127.0.0.1#011rip=127.0.0.1#011lport=143#011rport=52540#011resp=<hidden>
Mar 30 19:12:16 myvps dovecot: auth: Debug: ldap(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): Performing passdb lookup
Mar 30 19:12:16 myvps dovecot: auth: Debug: ldap(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): pass search: base=ou=people,dc=professional-eth0s,dc=ch scope=subtree filter=(&(objectClass=organizationalPerson)(uid=user)(mail=user@example.com)) fields=uid,userPassword
Mar 30 19:12:16 myvps dovecot: auth: Debug: ldap(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): result: uid=user userPassword=<hidden>; uid,userPassword unused
Mar 30 19:12:16 myvps dovecot: auth: Debug: ldap(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): username changed user@example.com -> user
Mar 30 19:12:16 myvps dovecot: auth: Debug: ldap(user,127.0.0.1,<9dBBsuQUPM1/AAAB>): username changed user -> user@example.com
Mar 30 19:12:16 myvps dovecot: auth: ldap(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): Password mismatch
Mar 30 19:12:16 myvps dovecot: auth: Debug: ldap(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): Finished passdb lookup
Mar 30 19:12:16 myvps dovecot: auth: Debug: oauth2(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): Performing passdb lookup
Mar 30 19:12:16 myvps dovecot: auth: Debug: oauth2(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): Making introspection request to https://authelia-roundcube-oidc:***@auth.example.com/api/oidc/introspection
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: Host created
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: Host session created
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: Need to perform DNS lookup
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: Performing asynchronous DNS lookup
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: conn unix:dns-client: dns(auth.example.com): Lookup started
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: conn unix:dns-client: Connecting
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: conn unix:dns-client (pid=507414,uid=0): Client connected (fd=22)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: conn unix:dns-client (pid=507414,uid=0): Sending version handshake
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: request [Req1: POST https://auth.example.com/api/oidc/introspection]: Submitted (requests left=1)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: conn unix:dns-client (pid=507414,uid=0): dns(auth.example.com): Lookup successful after 3 msecs
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: DNS lookup successful; got 2 IPs
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: peer [*redacted*]:443 (shared): Peer created
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: peer [*redacted*]:443: Peer pool created
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: peer [*redacted*]:443: Peer created
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: queue https://auth.example.com:443: Setting up connection to [*redacted*]:443 (SSL=auth.example.com) (1 requests pending)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: peer [*redacted*]:443: Linked queue https://auth.example.com:443 (1 queues linked)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: queue https://auth.example.com:443: Started new connection to [*redacted*]:443 (SSL=auth.example.com)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: conn unix:dns-client (pid=507414,uid=0): Disconnected: Connection closed (fd=22)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: conn unix:dns-client (pid=507414,uid=0): Disconnect: deinit
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: peer [*redacted*]:443: Creating 1 new connections to handle requests (already 0 usable, connecting to 0, closing 0)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: peer [*redacted*]:443: Making new connection 1 of 1 (0 connections exist, 0 pending)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: conn [*redacted*]:443 [1]: Connecting
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: conn [*redacted*]:443 [1]: Waiting for connect (fd=22) to finish for max 0 msecs
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: conn [*redacted*]:443 [1]: HTTPS connection created (1 parallel connections exist)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: conn [*redacted*]:443 [1]: Client connected (fd=22)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: conn [*redacted*]:443 [1]: Connected
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: conn [*redacted*]:443 [1]: Starting SSL handshake
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x10, ret=1: before SSL initialization
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: before SSL initialization
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSLv3/TLS write client hello
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1002, ret=-1: SSLv3/TLS write client hello
Mar 30 19:12:16 myvps dovecot: message repeated 2 times: [ auth: Debug: auth.example.com: SSL: where=0x1002, ret=-1: SSLv3/TLS write client hello]
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSLv3/TLS write client hello
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSLv3/TLS read server hello
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: TLSv1.3 read encrypted extensions
Mar 30 19:12:16 myvps dovecot: auth: Received valid SSL certificate: /C=US/O=Internet Security Research Group/CN=ISRG Root X1
Mar 30 19:12:16 myvps dovecot: auth: Received valid SSL certificate: /C=US/O=Let's Encrypt/CN=R3
Mar 30 19:12:16 myvps dovecot: auth: Received valid SSL certificate: /CN=example.com
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSLv3/TLS read server certificate
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: TLSv1.3 read server certificate verify
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSLv3/TLS read finished
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSLv3/TLS write change cipher spec
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSLv3/TLS write finished
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x20, ret=1: SSL negotiation finished successfully
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1002, ret=1: SSL negotiation finished successfully
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: conn [*redacted*]:443 [1]: SSL handshake successful
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: conn [*redacted*]:443 [1]: Ready for requests
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: peer [*redacted*]:443: Successfully connected (1 connections exist, 0 pending)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: peer [*redacted*]:443: Successfully connected (1 connections exist, 0 pending)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: peer [*redacted*]:443: Using 1 idle connections to handle 1 requests (1 total connections ready)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: queue https://auth.example.com:443: Connection to peer [*redacted*]:443 claimed request [Req1: POST https://auth.example.com/api/oidc/introspection]
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: conn [*redacted*]:443 [1]: Claimed request [Req1: POST https://auth.example.com/api/oidc/introspection]
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: request [Req1: POST https://auth.example.com/api/oidc/introspection]: Sent header
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: request [Req1: POST https://auth.example.com/api/oidc/introspection]: Send more (sent 105, buffered=0)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: request [Req1: POST https://auth.example.com/api/oidc/introspection]: Finished sending payload
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: peer [*redacted*]:443: No more requests to service for this peer (1 connections exist, 0 pending)
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSL negotiation finished successfully
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSL negotiation finished successfully
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSLv3/TLS read server session ticket
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1002, ret=1: SSL negotiation finished successfully
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSL negotiation finished successfully
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSL negotiation finished successfully
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1001, ret=1: SSLv3/TLS read server session ticket
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth.example.com: SSL: where=0x1002, ret=1: SSL negotiation finished successfully
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: conn [*redacted*]:443 [1]: Got 200 response for request [Req1: POST https://auth.example.com/api/oidc/introspection]: OK (took 96 ms + 32 ms in queue)
Mar 30 19:12:16 myvps dovecot: auth: Debug: oauth2(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): Introspection succeeded
Mar 30 19:12:16 myvps dovecot: auth: Debug: oauth2(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): Processing field active
Mar 30 19:12:16 myvps dovecot: auth: Debug: oauth2(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): Processing field client_id
Mar 30 19:12:16 myvps dovecot: auth: Debug: oauth2(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): Processing field iat
Mar 30 19:12:16 myvps dovecot: auth: Debug: oauth2(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): Processing field scope
Mar 30 19:12:16 myvps dovecot: auth: Debug: oauth2(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): oauth2 active_attribute is not configured; skipping the check
Mar 30 19:12:16 myvps dovecot: auth: Error: oauth2(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): oauth2 failed: Introspection failed: No username returned
Mar 30 19:12:16 myvps dovecot: auth: Debug: oauth2(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): Finished passdb lookup
Mar 30 19:12:16 myvps dovecot: auth: Debug: auth(user@example.com,127.0.0.1,<9dBBsuQUPM1/AAAB>): Auth request finished
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: conn [*redacted*]:443 [1]: Response payload stream destroyed (0 ms after initial response)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: request [Req1: POST https://auth.example.com/api/oidc/introspection]: Finished
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: queue https://auth.example.com:443: Dropping request [Req1: POST https://auth.example.com/api/oidc/introspection]
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: host auth.example.com: Host is idle (timeout = 1799900 msecs)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: request [Req1: POST https://auth.example.com/api/oidc/introspection]: Free (requests left=1)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client[1]: peer [*redacted*]:443: No requests to service for this peer (1 connections exist, 0 pending)
Mar 30 19:12:16 myvps dovecot: auth: Debug: http-client: conn [*redacted*]:443 [1]: No more requests queued; going idle (timeout = 60000 msecs)
Mar 30 19:12:18 myvps dovecot: auth: Debug: client passdb out: FAIL#0111#011user=user@example.com#011code=temp_fail#011resp=eyJzdGF0dXMiOiI0MDEiLCJzY2hlbWVzIjoiYmVhcmVyIiwic2NvcGUiOiJtYWlsIn0=
Mar 30 19:12:18 myvps dovecot: imap-login: Disconnected: Connection closed (client didn't finish SASL auth, waited 2 secs): user=<user@example.com>, method=XOAUTH2, rip=127.0.0.1, lip=127.0.0.1, secured, session=<9dBBsuQUPM1/AAAB>
```

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

The issue you're encountering with Dovecot complaining about "oauth2 failed: Introspection failed: No username returned" after upgrading Authelia to v4.38.6 seems to be related to the introspection response not containing a username attribute that Dovecot expects. This is indicated by the error message in the Dovecot logs.

In the provided Dovecot configuration (`dovecot-oauth2.conf.ext`), you have specified `username_attribute = username`. This means Dovecot expects the introspection response from Authelia to contain a `username` attribute that it can use to identify the user. However, the logs suggest that this attribute is either not present or not correctly populated in the introspection response from Authelia.

To resolve this issue, you need to ensure that Authelia's introspection response includes the `username` attribute with the correct value for Dovecot to use. This might involve adjusting the Authelia configuration to ensure that user information, including the username, is correctly included in the introspection response.

Given the information provided, here are some steps you can take:

1. **Review Authelia OIDC Configuration**: Ensure that Authelia's OIDC configuration is set up to include user information in the token and introspection responses. This might involve configuring user claims or scopes correctly in Authelia to ensure that the `username` attribute is included.

2. **Check Authelia Version Release Notes**: Since the issue appeared after upgrading Authelia, review the release notes for Authelia v4.38.6 and any versions in between from your previous version. Look for any changes related to OIDC, introspection responses, or user information handling that might require adjustments in your configuration.

3. **Debugging and Logging**: Increase the logging level for both Authelia and Dovecot to capture more detailed information about the introspection request and response. This can help identify what attributes are being returned by Authelia and what might be missing or misconfigured.

4. **Consult Authelia Documentation**: Review the latest Authelia documentation for any guidance on configuring OIDC and introspection responses. There might be new configuration options or recommendations in the latest version that you need to apply.

5. **Community and Support Channels**: If the issue persists, consider reaching out to the Authelia community or support channels. Other users or the developers might have encountered similar issues and can offer specific advice or solutions.

Unfortunately, without direct access to the introspection response from Authelia, it's challenging to provide a more specific solution. The steps above are aimed at guiding you through troubleshooting and identifying the configuration adjustments needed to resolve the issue.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment, marked as answer]
Think this is a regression, give `authelia/authelia:fix-introspection-claims` a try.

**Reply by sphakka** [reply]
Yeah, Authelia v4.38.7 resolves the issue, thanks!

**Comment by wawefr** [comment]
Hello I'm on version v4.38.8 and I have the same problem ?  (@sphakka it's now ok for you?)
Thanks for your help

**Comment by sphakka** [comment]
@wawefr, that was fixed for me with v4.38.7. It's been working with versions v4.38.8 and latest v4.38.9 -- all Debian builds.

**Reply by wawefr** [reply]
@sphakka  thanks for your feedback. I share my logs perhaps you can help me a little. Thx

For your information the user frederic@****.club is the "same" in dovecot/roundcube and Authelia with same pwd under Argon2I 
I use the same parameters than you in Dovecot/roundcube and Authelia

### **This is the log Dovecot-info.log**

Jul 26 18:45:13 auth-worker(218285): Info: conn unix:auth-worker (pid=218276,uid=112): auth-worker<1>: sql(frederic@*****.club,68.66.251.121,<iYk8PCkeNohEQvt5>): Password mismatch
Jul 26 18:45:15 imap-login: Info: Disconnected: Connection closed (auth service reported temporary failure): user=<frederic@*****.club>, method=XOAUTH2, rip=68.66.251.121, lip=68.66.251.121, TLS, session=<iYk8PCkeNohEQvt5>
### **This is the log dovecot.log**

Jul 26 18:45:13 auth: Error: oauth2(frederic@******.club,68.66.251.121,<iYk8PCkeNohEQvt5>): oauth2 failed: Introspection failed: No username returned
### ** This is the log Roundcube**

[26-Jul-2024 16:52:57 +0000]: <lv7r32bs> PHP Error: OAuth token request failed: Client error: `POST https://authv2.******.online/api/oidc/token` resulted in a `400 Bad Request` response:
{"error":"invalid_grant","error_description":"The provided authorization grant (e.g., authorization code, resource owner (truncated...)
; mail.patat-monnot.club GuzzleHttp/7 - [26/Jul/2024:16:52:57 +0000] "POST /api/oidc/token HTTP/1.1" 400 331 in /usr/share/roundcube/program/include/rcmail_oauth.php on line 324 (GET /index.php/login/oauth?code=authelia_ac_JIYrNi1tRhRwzg0_wHp3vbJXBMud5COn-H8oxt_D63I.S8D3ifqcgbFirJb6nGrYfbuTT3r8SmJLrrGmRhku9IE&iss=https%3A%2F%2Fauthv2.******.online&scope=email>
[26-Jul-2024 16:53:13 +0000]: <lv7r32bs> IMAP Error: Login failed for frederic@******.club against mail.******.club from 92.169.69.39. AUTHENTICATE XOAUTH2: A0002 NO [UNAVAILABLE] Temporary authentication failure. [mail.********.club:2024-07-26 16:53:13] in /usr/share/roundcube/program/lib/Roundcube/rcube_imap.php on line 211 (GET /index.php/login/o>

**Comment by sphakka** [comment]
@wawefr, you're welcome! I'm afraid I can't do much, as am just a regular user -- hopefully, some dev will jump in. 

What's your app stack version? Mine on Debian: Roundcube v1.6.7+dfsg-1, Dovecot v1:2.3.21+dfsg1-2 (basically 2.3.21, not the latest patch level though), Authelia v4.38.9-1, Nginx v1.22.1-9.
What's you're authentication backend? I'm on LDAP: slapd v2.5.13+dfsg-5.
Are you able to login to Roundcube via Authelia OIDC and see the GUI mail view? That was my case, indeed, but couldn't connect to Dovecot/IMAP to read/send e-mails. I see you get an `OAuth token request failed` which didn't occur to me and seems to come from an upstream auth problem, i.e., not Dovecot's fault.
Are you able to login to Dovecot/IMAP by other means, e.g., mail client via LDAP/DB auth? 

Meanwhile, please, have also a look at this [integration doc](https://github.com/authelia/authelia/blob/master/docs/content/integration/openid-connect/roundcube/index.md) (still not published on web site, weird...). Check your config against that (mine is still as in my original post above), enable debugging everywhere and report:

* Dovecot > `dovecot-oauth2.conf.ext`: `debug = yes`
* Roundcube > `config.inc.php`: `$config['imap_debug'] = true;`
* Authelia > `configuration.yml`: `enable_client_debug_messages: true

HTH!

**Comment by wawefr** [comment]
@sphakka  Many thanks and I hope you are well.

My configuration Dovecot v2.3.19.1/ Roundcube 1.6.7/ Authelia v4.38.8
I'm using Cloudflare tunnel (more secure for self-hosting, I prefer than NGINX)  

Roundcube and Dovecot work perfectly with login/password under Roundcube. 
When I use authentication via Authelia, everything works well under Authelia, and I get back to the Roundcube page, but with the message 'Unable to connect to the server (Dovecot)'."

This indicates that while Roundcube and Dovecot function properly with direct login, there's an issue with connecting to Dovecot when Authelia is involved in the authentication process. (Authelia log confirm) 

When I look at the Dovecot logs, it receives the username from Authelia (I connect via email), but Dovecot cannot find it (my Dovecot database only has the email addresses, not the usernames). I think this might be the issue. How do you manage users in Dovecot? 
![image](https://github.com/user-attachments/assets/9d9ce08e-16ac-40e9-bb4b-1a5336042ff9)


Have a good day Fred

**Comment by sphakka** [comment]
With Dovecot v2.3.19 I couldn't get introspection to work, but [something was fixed in v2.3.21](https://dovecot.org/doc/NEWS-2.3): you should upgrade then.

My auth backend, as said before, is LDAP: users authenticate with their e-mail address, however in Dovecot Oauth2 (introspection with "post" method as stated above in my original post) I had to use "username" as `username_attribute`. 

Indeed, from your your logs, introspection works but fails because Dovecot receives an e-mail address when it should be a simple user name. Thus you may try (see [other examples here](https://doc.dovecot.org/configuration_manual/authentication/oauth2/)):
```
username_attribute = email
```

Otherwise the issue might be with your SQL auth backend (maybe an incomplete `user_query`). 

If that doesn't work, please, 
* open a **new discussion**, because I really think it's just a configuration problem, unrelated to Authelia 4.38.6's bug,
* upgrade your stack to the latest versions available,
* post your relevant configuration for the whole stack and the log snippets **properly formatted as code blocks**.

**Comment by wawefr** [comment]
Hi @sphakka  many thanks for your help.  I agree with your analysis. I will look into it and get back to you with an update. Have a good day Fred
